File: kiosk/src/lib/heartbeat.ts
/**
 * While someone is signed in, tell the API this kiosk is alive: an
 * immediate beat, then one every minute. Each beat upserts the kiosk's
 * Device row and returns the registration state for the shell chip.
 * Failures (offline, read-only mode's 423, a 403) keep the last state
 * and are retried on the next tick.
 */

import { heartbeatRequest, type RegistrationState } from './api';
import { kioskVersion } from './config';
import { getIdentity } from './identity';
import { platform } from './platform';

export const HEARTBEAT_MS = 60_000;

export interface HeartbeatHandle {
  stop(): void;
  /** Beat right now (after a rename). Resolves after the attempt. */
  now(): Promise<void>;
}

export function startHeartbeat(
  onState: (state: RegistrationState) => void,
  intervalMs: number = HEARTBEAT_MS,
  signIn?: { method: 'password' | 'link' },
): HeartbeatHandle {
  let stopped = false;
  // Seeded from the sign-in that started this heartbeat; every beat sends
  // sign_in while this is set, and it's cleared only once a beat carrying
  // it RESOLVES successfully — a failed beat leaves it set so the next
  // tick (or now()) retries until the server has recorded the sign-in.
  let pendingSignIn = signIn;
  const beat = async () => {
    if (stopped) return;
    const { serial, name } = getIdentity();
    const asSignIn = pendingSignIn;
    try {
      const result = await heartbeatRequest({
        serial, name, mode: platform().mode, version: kioskVersion(),
        ...(asSignIn ? { sign_in: true, login_method: asSignIn.method } : {}),
      });
      if (asSignIn) pendingSignIn = undefined;
      if (!stopped) onState(result.registration);
    } catch {
      /* keep the last known state (and any pending sign-in); next tick retries */
    }
  };
  void beat();
  const timer = setInterval(() => void beat(), intervalMs);
  return {
    stop() {
      stopped = true;
      clearInterval(timer);
    },
    now: () => beat(),
  };
}

File: Android_Kiosk_App/app/src/main/java/com/serversherpa/kiosk/data/heartbeat/Heartbeat.kt
package com.serversherpa.kiosk.data.heartbeat

import com.serversherpa.kiosk.core.devices.RegistrationState
import com.serversherpa.kiosk.core.model.HeartbeatIn
import com.serversherpa.kiosk.data.api.KioskApi
import com.serversherpa.kiosk.data.auth.LoginMethod
import com.serversherpa.kiosk.data.config.KioskConfig
import com.serversherpa.kiosk.data.identity.Identity
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.CoroutineStart
import kotlinx.coroutines.Job
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.isActive
import kotlinx.coroutines.launch

/**
 * While someone is signed in and the app is in front: an immediate beat,
 * then one a minute. Each beat upserts the kiosk's Device row and returns
 * the registration state. Failures keep the last state (and any pending
 * sign_in) and are retried next tick.
 */
class Heartbeat(
    private val api: KioskApi,
    private val identity: Identity,
    private val config: KioskConfig,
    private val deviceInfo: () -> Map<String, String>,
    private val intervalMs: Long = 60_000,
) {
    private val _registration = MutableStateFlow<RegistrationState?>(null)
    val registration: StateFlow<RegistrationState?> = _registration

    private var job: Job? = null
    @Volatile private var pendingSignIn: LoginMethod? = null

    fun start(scope: CoroutineScope, signIn: LoginMethod?) {
        stop()
        if (signIn != null) pendingSignIn = signIn
        job = scope.launch(start = CoroutineStart.UNDISPATCHED) {
            while (isActive) {
                beat()
                delay(intervalMs)
            }
        }
    }

    fun stop() {
        job?.cancel(); job = null
        _registration.value = null
    }

    /** Beat right now (after a rename). Resolves after the attempt. */
    suspend fun now() = beat()

    private suspend fun beat() {
        val (serial, name) = identity.get()
        val asSignIn = pendingSignIn
        try {
            val result = api.heartbeat(HeartbeatIn(
                serial = serial, name = name, mode = "android", version = config.kioskVersion,
                raw_info = deviceInfo(), sign_in = asSignIn != null, login_method = asSignIn?.wire,
            ))
            if (asSignIn != null && pendingSignIn === asSignIn) pendingSignIn = null
            _registration.value = RegistrationState.fromWire(result.registration)
        } catch (e: Exception) {
            /* keep the last known state and any pending sign-in */
        }
    }
}
